Documentation/contributing/development/introducing_new_crds.rst

```diff
@@ -0,0 +1,400 @@
+Introducing new CRDs
+====================
+
+Cilium uses a combination of code generation tools to facilitate adding
+CRDs to the Kubernetes instance it is installed on.
+
+These CRDs make themselves available in the generated Kubernetes client
+Cilium uses.
+
+Defining And Generating CRDs
+----------------------------
+
+Paths
+
+::
+
+   pkg/k8s/apis/cilium.io/v2/
+   pkg/k8s/apis/cilium.io/v2alpha1/
+
+Currently, two API versions exist ``v2`` and ``v2alpha1``.
+
+CRDs are defined via Golang structures, annotated with ``marks``, and
+generated with Cilium make file targets.
+
+Marks
+~~~~~
+
+Marks are used to tell ``controller-gen`` *how* to generate the CRD.
+This includes defining the CRD’s various names (Singular, plural,
+group), it’s Scope (Cluster, Namespaced), Shortnames, etc…
+
+An example:
+
+::
+
+   // +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+   // +kubebuilder:resource:categories={cilium},singular="ciliumendpointslice",path="ciliumendpointslices",scope="Cluster",shortName={ces}
+
+   // +kubebuilder:storageversion
+
+You can find CRD generation ``marks`` documentation
+`here <https://book.kubebuilder.io/reference/markers/crd.html>`__.
+
+Marks are also used to generate json-schema validation. You can define
+validation criteria such as “format=cidr” and “required” via validation
+``marks`` in your struct’s comments.
+
+An example:
+
+.. code:: go
+
+   type CiliumBGPPeeringConfiguration struct {
+       // PeerAddress is the IP address of the peer.
+       // This must be in CIDR notation and use a /32 to express
+       // a single host.
+       //
+       // +kubebuilder:validation:Required
+       // +kubebuilder:validation:Format=cidr
+       PeerAddress string `json:"peerAddress"`
+
+You can find CRD validation ``marks`` documentation
+`here <https://book.kubebuilder.io/reference/markers/crd-validation.html>`__.
+
+Defining CRDs
```

Comment:
Previous same-level title was “Defining And Generating CRDs”, this probably needs some renaming somewhere to avoid defining the CRDs in two different sections?

Reply:
Sorry not following. This heading is not the same as "Defining And Generating CRDS". What would you like changed?

Reply:
What I meant is that the hierarchy of your document is like this:

```
Defining And Generating CRDs
----------------------------

<section 1 ...>

Defining CRDs
-------------

<section 2...>
```

The two sections mention “defining CRDs”. So if I want to read about defining CRDs, should I read the first or the second one? It would sound more intuitive to have either something like this:

```
Defining And Generating CRDs
----------------------------

[...]

Defining CRDs
~~~~~~~~~~~~~

[...]

Generating CRDs
~~~~~~~~~~~~~~~

[...]
```

where “Defining CRDs” is a subsection of “Defining And Generating CRDs”, or something like this:

```
Defining And Generating CRDs
----------------------------

[...]

Something Else About CRDs, That Supposes They Have Been Defined Already
-----------------------------------------------------------------------

[...]
```

where we don't hint about defining them in two different follow-up sections.

Does it make sense?

Reply:
```
Defining And Generating CRDs
----------------------------

[...]

Defining CRDs
~~~~~~~~~~~~~

[...]

Generating CRDs
~~~~~~~~~~~~~~~

[...]
```

Ahh I see now, the above heading levels make the most sense to me. I'll update.